// castv2-client/lib/controllers/connection.js
var util            = require('util');
var debug           = require('debug')('castv2-client');
var JsonController  = require('./json');

function ConnectionController(client, sourceId, destinationId) {
  console.log('CONSTRUCTOR ConnectionController');
  JsonController.call(this, client, sourceId, destinationId, 'urn:x-cast:com.google.cast.tp.connection');

  var self = this;
  var connected = true;
  this.sourceId = sourceId;

  function onmessage(data, broadcast) {
    console.log('ON MESSAGE DATA TYPE', data.type);
    if(data.type === 'CLOSE') {
      connected = false;
      self.emit('disconnect');
    }
  }

  function onclose() {
    console.log('connection close', sourceId);
    self.removeListener('message', onmessage);
    if(connected) {
      console.log('connection force disconnect', sourceId);
      self.emit('disconnect');
    }
  }

  this.on('message', onmessage);
  this.once('close', onclose);
}

util.inherits(ConnectionController, JsonController);

ConnectionController.prototype.connect = function() {
  console.log('connect called', this.sourceId);
  this.emit('test');
  this.send({ type: 'CONNECT' });
};

ConnectionController.prototype.disconnect = function() {
  this.send({ type: 'CLOSE' });
};

module.exports = ConnectionController;
